pkg/session/pipelineddmltest/pipelineddml_test.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pipelineddmltest
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb/pkg/testkit"
+	"github.com/stretchr/testify/require"
+)
+
+func TestVariable(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	require.Equal(t, tk.Session().GetSessionVars().BulkDMLEnabled, false)
+	tk.MustExec("set session tidb_dml_type = bulk")
+	require.Equal(t, tk.Session().GetSessionVars().BulkDMLEnabled, true)
+	tk.MustExec("set session tidb_dml_type = standard")
+	require.Equal(t, tk.Session().GetSessionVars().BulkDMLEnabled, false)
+	// not supported yet.
+	tk.MustExecToErr("set session tidb_dml_type = bulk(10)")
+}
+
+func compareTables(t *testing.T, tk *testkit.TestKit, t1, t2 string) {
+	t1Rows := tk.MustQuery("select * from " + t1).Sort().Rows()
+	t2Rows := tk.MustQuery("select * from " + t2)
+	require.Equal(t, len(t1Rows), len(t2Rows.Rows()))
+	t2Rows.Sort().Check(t1Rows)
+}
+
+func prepareData(tk *testkit.TestKit) {
+	tk.MustExec("drop table if exists t, _t")
+	tk.MustExec("create table t (a int primary key, b int)")
+	tk.MustExec("create table _t like t")
+	results := make([]string, 0, 10000)
+	for i := 0; i < 10000; i++ {
+		tk.MustExec("insert into t values (?, ?)", i, i)
+		results = append(results, fmt.Sprintf("%d %d", i, i))
+	}
+	tk.MustQuery("select * from t order by a asc").Check(testkit.Rows(results...))
+}
+
+func TestPipelinedDMLInsert(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	prepareData(tk)
+	tk.MustExec("set session tidb_dml_type = bulk")
```

Comment:
Could you add a test that without pipelined DML, the following bulk insert will fail cause by exceeded memory limit?

Reply:
I added it in the end of this file, note the mem quota is carefully choosed because when mutations are not such large(reach GB level), memory used by read executors can be large also.